src/backends/x11/meta-monitor-manager-xrandr.c

```diff
@@ -2091,48 +2091,149 @@ meta_monitor_manager_xrandr_class_init (MetaMonitorManagerXrandrClass *klass)
     g_quark_from_static_string ("-meta-monitor-xrandr-data");
 }
 
+static gboolean
+meta_monitor_manager_xrandr_update_underscan_crop (MetaMonitorManagerXrandr *manager_xrandr)
```

Comment:
Man, this code is hairy. I didn't realize that `crop` mode was something different than what upstream was doing. We need to get away from that ASAP.

Reply:
AFAIK, this is done this way because of a limitation on the Intel driver, that can't do the right thing in HW and needs to do this "cropping" in software.

But yes, it's insane, it does not work with all resolutions (although for the hardcoded set we consider in this PR it seems to work) and seems to mess up very nicely with the configuration history that mutters holds in MetaConfigManager.

Reply:
I actually think it's the other way around (wish I knew who the right person to confirm is). I think normally what happens is that you tell the driver you want to underscan and it simply scales down the image and puts a black border around it. That (I believe) could happen entirely in software or maybe some hardware has the ability to simply scale the image itself.

What the intel driver does is much heavier. It actually changes the available modes so each of them is smaller and "skewed" (not really sure what that means) and requires you to set the mode. So, you're actually setting a new mode on the CRTC. If the mutter code works in the `on` case, then I have to believe that other drivers simply scale without changing the mode. One of the things I was confused about when reading this again is why it wasn't implemented as more or less an alias to the randr scale option.

For the configuration history, one thing I noticed when reading this function is that it's the only place here where you create a `MetaOutputInfo` based on a `MetaOutput`. The rest of the time, the `MetaOutputInfo` is always created in `meta_monitor_config_manager_assign`. I have a feeling that the on the fly creation of the `MetaOutputInfo`s and `MetaCRTCInfo`s that are being passed to `apply_crtc_assignments` has something to do with the breakage.